src/prefect/engine/task_runner.py

```diff
@@ -93,7 +93,10 @@ def __init__(
         else:
             self.result = Result().copy() if result is None else result.copy()
             if self.task.target:
-                self.result.location = self.task.target
+                if not isinstance(self.task.target, str):
+                    self.result.location = self.task.target(**prefect.context)
```

Comment:
So I think this logic needs to be moved somewhere else deeper in the pipeline; e.g., if I used `lambda **kwargs: kwargs['task_run_count']` that wouldn't work here because `task_run_count` isn't populated until we're within the pipeline.